django/db/backends/postgresql/schema.py

```diff
@@ -42,7 +42,10 @@ def _field_indexes_sql(self, model, field):
     def _field_data_type(self, field):
         if field.is_relation:
             return field.rel_db_type(self.connection)
-        return self.connection.data_types[field.get_internal_type()]
+        return self.connection.data_types.get(
+            field.get_internal_type(),
+            field.db_type(self.connection),
+        )
```

Comment:
Actually `return field.db_type(self.connection)` should be enough as it defaults to `connection.data_types.get(self.get_internal_type())`

https://github.com/django/django/blob/ff5dfbc63a278219cd929449678b99ebec9a4b5f/django/db/models/fields/__init__.py#L664-L688

Reply:
Unfortunately we cannot use it because it fills parameters what will brake the original fix, e.g. `connection.data_types['CharField']` is always `varchar(%(max_length)s)'` regardless of precision and we want to cast values only when data types changed (see [postgresql/schema.py#L71-L73](https://github.com/django/django/blob/master/django/db/backends/postgresql/schema.py#L71-L73)).

Reply:
Makes sense.

Falling back to `field.db_type(self.connection)` will cause silent truncation of `ArrayField(CharField(max_length=100)) -> ArrayField(CharField(max_length=50))` and `ArrayField(Field(), size=10) -> ArrayField(Field(), size=5)` though but I guess it's better than crashing.

Reply:
Yes :+1: I'm going to reopen [the original ticket](https://code.djangoproject.com/ticket/28816) with this use case.